### DISCUSSION TITLE
Addon Settings Service

### DISCUSSION THREAD

**Discussion by blairliikala** [ExpressionEngine/ExpressionEngine#2863] [Feature Requests]
Just about every module needs settings, and just about every module needs to re-invent the wheel to build settings (or copy from their last add-on).  In the effort of better standardizing boilerplate, having a pre-made Settings service or having premade settings as part of the add-on builder, would be extremely helpful for spinning up new add-ons, and for new to the platform developers building their first add-on.

I could see this in a few different capacities:
- Include methods to add, replace, remove, and import/export settings. ..etc.
- Methods to include settings in other services and classes (with special care to fieldtypes that already have `$this->settings`)
- Cascade order for getting settings: default config file -> config.php -> env files -> database.  If the setting doesn't exist in the database, then check the env file, then config file...etc.
- Options to remove unused settings in the database.
- Encrypting into the database, useful for API keys.
- Creating the database table and removing.
- Settings validation (though might be OK with the existing validation service?)

If this were offered in the builder then it could provide some good training-by-example for how services and models work.

If more complex settings were needed then the developer would either extend this or build their own implementation.

Existing add-ons moving to this system could use a method or import process for existing settings.

**Comment by intoeetive** [comment]
@TomJaeger FYI
